From c9d872c876e3e33a8768a9a00676f014e7d75c61 Mon Sep 17 00:00:00 2001
From: Marius Gripsgard <marius@ubports.com>
Date: Thu, 3 Oct 2019 01:56:54 +0200
Subject: Use drmModeGetResources as fallback for modeset probe

This allows the use of drmModeGetResources to check if KMS is supported
on devices where drmCheckModesettingSupported is not supported or fails
to detect.

Since we use drmModeGetResources later on, it's a safe check since if
this fails, we will fail later anyway.

Origin: vendor
Forwarded: not-needed
Last-Update: 2021-06-16
---
 src/platforms/mesa/server/display_helpers.cpp | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/platforms/mesa/server/display_helpers.cpp b/src/platforms/mesa/server/display_helpers.cpp
index b357f6c67b..c2b09a9542 100644
--- a/src/platforms/mesa/server/display_helpers.cpp
+++ b/src/platforms/mesa/server/display_helpers.cpp
@@ -112,6 +112,7 @@ mgmh::DRMHelper::open_all_devices(
         }
         else
         {
+            drmModeRes* drmDev;
             switch (auto err = -drmCheckModesettingSupported(busid.get()))
             {
             case 0: break;
@@ -127,9 +128,18 @@ mgmh::DRMHelper::open_all_devices(
                 mir::log_debug("MIR_MESA_KMS_DISABLE_MODESET_PROBE is set");
                 // Falls through.
             case EINVAL:
+                // Trying fallback
                 mir::log_warning(
-                    "Failed to detect whether device %s supports KMS, but continuing anyway",
+                    "Failed to detect whether device %s supports KMS, checking using fallback method",
                     device.devnode());
+
+                drmDev = drmModeGetResources(tmp_fd);
+                if (!drmDev && getenv("MIR_MESA_KMS_DISABLE_MODESET_PROBE") == nullptr) {
+                    mir::log_info("Ignoring non-KMS DRM device %s", device.devnode());
+                    drmModeFreeResources(drmDev);
+                    continue;
+                }
+                drmModeFreeResources(drmDev);
                 break;
 
             default:
-- 
2.25.1

